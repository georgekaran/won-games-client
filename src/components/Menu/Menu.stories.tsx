import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import Menu from './Menu'

export default {
  title: 'Menu',
  component: Menu
} as Meta

export const Default: Story = () => <Menu />

Default.parameters = {
  layout: 'fullscreen',
  backgrounds: {
    default: 'dark'
  }
}
