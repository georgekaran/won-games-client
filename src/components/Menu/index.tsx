import React, { useState } from 'react'
import { Menu2 as MenuIcon } from '@styled-icons/remix-line/Menu2'
import { Search as SearchIcon } from '@styled-icons/material-outlined/Search'
import { ShoppingCart } from '@styled-icons/material-outlined/ShoppingCart'
import { Close as CloseIcon } from '@styled-icons/material-outlined/Close'

import * as S from './styles'
import Logo from '@/components/Logo'
import Button from '@/components/Button'
import MediaMatch from '@/components/MediaMatch'

export type MenuProps = {
  username?: string
}

const Menu = ({ username }: MenuProps) => {
  const [isOpen, setIsOpen] = useState(false)
  return (
    <S.Wrapper>
      <MediaMatch data-testid="media-menu-icon" lessThan="medium">
        <S.IconWrapper>
          <MenuIcon aria-label="Open menu" onClick={() => setIsOpen(true)} />
        </S.IconWrapper>
      </MediaMatch>

      <S.LogoWrapper>
        <Logo hideOnMobile />
      </S.LogoWrapper>

      <MediaMatch greaterThan="medium">
        <S.MenuNav>
          <S.MenuLink href="#">Home</S.MenuLink>
          <S.MenuLink href="#">Explore</S.MenuLink>
        </S.MenuNav>
      </MediaMatch>

      <S.MenuGroup>
        <S.IconWrapper>
          <SearchIcon aria-label="Search" />
        </S.IconWrapper>
        <S.IconWrapper>
          <ShoppingCart aria-label="Open shopping cart" />
        </S.IconWrapper>
        <MediaMatch greaterThan="medium">
          {!username && (
            <Button>
              Sign in
            </Button>
          )}
        </MediaMatch>
      </S.MenuGroup>

      <S.MenuFull aria-hidden={!isOpen} isOpen={isOpen}>
        <CloseIcon aria-label="Close menu" onClick={() => setIsOpen(false)} />
        <S.MenuNav>
          <S.MenuLink href="#">Home</S.MenuLink>
          <S.MenuLink href="#">Explore</S.MenuLink>
          {!!username && (
            <>
              <S.MenuLink href="#">My account</S.MenuLink>
              <S.MenuLink href="#">Wishlist</S.MenuLink>
            </>
          )}
        </S.MenuNav>

        {!username && (
          <S.RegisterBox>
            <Button fullWidth size="large">
              Log in now
            </Button>
            <span>or</span>
            <S.CreateAccount href="#" title="Sign up">
              Sign up
            </S.CreateAccount>
          </S.RegisterBox>
        )}
      </S.MenuFull>
    </S.Wrapper>
  )
}

export default Menu
