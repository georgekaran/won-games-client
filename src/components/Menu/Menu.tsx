import React, { useState } from 'react'
import { Menu2 as MenuIcon } from '@styled-icons/remix-line/Menu2'
import { Search as SearchIcon } from '@styled-icons/material-outlined/Search'
import { ShoppingCart } from '@styled-icons/material-outlined/ShoppingCart'
import { Close as CloseIcon } from '@styled-icons/material-outlined/Close'

import * as S from './Menu.styles'
import Logo from '@/components/Logo/Logo'

export type MenuProps = {}

const Menu = () => {
  const [isOpen, setIsOpen] = useState(false)
  return (
    <S.Wrapper>
      <S.IconWrapper>
        <MenuIcon aria-label="Open menu" onClick={() => setIsOpen(true)} />
      </S.IconWrapper>
      <S.LogoWrapper>
        <Logo hideOnMobile />
      </S.LogoWrapper>
      <S.MenuGroup>
        <S.IconWrapper>
          <SearchIcon aria-label="Search" />
        </S.IconWrapper>
        <S.IconWrapper>
          <ShoppingCart aria-label="Open shopping cart" />
        </S.IconWrapper>
      </S.MenuGroup>

      <S.MenuFull aria-hidden={!isOpen} isOpen={isOpen}>
        <CloseIcon aria-label="Close menu" onClick={() => setIsOpen(false)} />
      </S.MenuFull>
    </S.Wrapper>
  )
}

export default Menu
