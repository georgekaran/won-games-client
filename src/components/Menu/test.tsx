import React from 'react'
import faker from 'faker'
import { screen, RenderResult, fireEvent } from '@testing-library/react'

import Menu, { MenuProps } from '.'
import { renderWithTheme } from '@/test/helpers'

const makeSut = ({ username = '' }: MenuProps = {}): RenderResult => {
  return renderWithTheme(<Menu username={username} />)
}

describe('<Menu />', () => {
  test('should render with initial state', () => {
    const { container } = makeSut()
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.getByLabelText(/open menu/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/search/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/open shopping cart/i)).toBeInTheDocument()
    expect(screen.getByTestId('logo')).toBeInTheDocument()
  })

  test('should handle the open/close mobile menu', () => {
    makeSut()
    // Select the menu
    const menuFull = screen.getByRole('navigation', { hidden: true })
    // Check if menu is hide
    expect(menuFull.getAttribute('aria-hidden')).toBe('true')
    expect(menuFull).toHaveStyleRule('opacity', '0')
    // Click in the button to open the menu e check if the menu is open
    const menuIcon = screen.getByLabelText(/open menu/i)
    fireEvent.click(menuIcon)
    expect(menuFull.getAttribute('aria-hidden')).toBe('false')
    expect(menuFull).toHaveStyleRule('opacity', '1')
    // Click in the button to close the menu e check if the menu is hide/close
    const closeMenuIcon = screen.getByLabelText(/close menu/i)
    fireEvent.click(closeMenuIcon)
    expect(menuFull.getAttribute('aria-hidden')).toBe('true')
    expect(menuFull).toHaveStyleRule('opacity', '0')
  })

  test('Should show My account and wishlist links if username is not nullish', () => {
    makeSut({ username: faker.internet.userName() })
    expect(screen.getByText(/my account/i)).toBeInTheDocument()
    expect(screen.getByText(/wishlist/i)).toBeInTheDocument()
  })

  test('Should not show RegisterBox if username is not nullish', () => {
    makeSut({ username: faker.internet.userName() })
    expect(screen.queryByText(/Log in now/i)).not.toBeInTheDocument()
    expect(screen.queryByText(/Sign up/i)).not.toBeInTheDocument()
  })
})
