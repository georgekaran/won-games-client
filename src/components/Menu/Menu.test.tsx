import React from 'react'
import { render } from '@testing-library/react'

import Menu from './Menu'

describe('<Menu />', () => {
  test('should render with initial state', () => {
    const { container } = render(<Menu />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
