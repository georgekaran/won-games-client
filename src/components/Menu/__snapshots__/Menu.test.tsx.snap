// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<Menu /> should render with initial state 1'] = `
<div
  class=""
>
  <h1>
    Menu
  </h1>
</div>
`
